test/test_ops.py

```diff
@@ -463,6 +439,115 @@ def _fn(*sample_args, **sample_kwargs):
+
+        # Short-circuits if output is not a single tensor
+        # TODO: extend this to handle multiple tensors
+        if not isinstance(expected, torch.Tensor):
+            self.skipTest("Skipped! Only supports single tensor outputs for now.")
+
+        # Case 0: out= with the correct shape, dtype, and device
+        #   but NaN values
+        #   Expected behavior: out= values have no effect on the computation.
+        out = torch.full_like(expected, float('nan'))
+        op_out(out=out)
+        self.assertEqual(expected, out)
+
+        # Case 1: out= with the correct shape, dtype, and device,
+        #   but noncontiguous.
+        #   Expected behavior: strides are respected.
+        # TODO: (implement) and switch to make_tensor_like()
+        out = make_tensor(expected.shape,
+                          dtype=expected.dtype,
+                          device=expected.device,
+                          discontiguous=True)
+        original_strides = out.stride()
+        op_out(out=out)
+        self.assertEqual(expected, out)
+        self.assertEqual(original_strides, out.stride())
+
+        # Case 2: out= with the correct dtype and device, but the wrong shape
+        #   Expected behavior: resize with a warning.
+        wrong_shape = list(expected.shape)
+        wrong_shape[-1] = wrong_shape[-1] + 1
+        out = make_tensor(wrong_shape,
+                          dtype=expected.dtype,
+                          device=expected.device)
+        with self.assertWarnsRegex(UserWarning, "An output with one or more elements"):
+            op_out(out=out)
+        self.assertEqual(expected, out)
+
+        # Case 3: out= with the correct dtype and device, but an empty
+        #   tensor.
+        #   Expected behavior: resize without warning.
+        # TODO: this currently asserts the op throws NO warnings,
+        #   which may be too strict (if the op is deprecated, for example)
+        out = make_tensor((0,),
+                          dtype=expected.dtype,
+                          device=expected.device)
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            op_out(out=out)
+        self.assertEqual(len(w), 0)
+        self.assertEqual(expected, out)
+
+        # Case 4: out= with correct shape and dtype, but wrong device.
+        #   NOTE: this case only tested if a CUDA device is available.
+        #   Expected behavior: error.
+        if torch.device(device).type == 'cuda':
+            # Case 4A: CUDA input with CPU out.
+            out = make_tensor(expected.shape,
+                              dtype=expected.dtype,
+                              device='cpu')
+            with self.assertRaises(RuntimeError):
+                op_out(out=out)
+
+            # Case 4B: CPU input with CUDA out.
+            #   TODO: not implemented.
+
+        # Case 5: out= with correct shape and device, but a dtype
```

Comment:
Add NOTE that enabling this requires further review.

Aside: should we limit type promotion to unary ufuncs, binary ufuncs, and reductions?